usr/src/man/man1/tip.1
'\" te
.\"  Copyright (c) 1980 Regents of the University of California.  All rights reserved.  The Berkeley software License Agreement  specifies the terms and conditions for redistribution.  Copyright (c) 2001, Sun Microsystems, Inc.  All Rights Reserved
.TH TIP 1 "Nov 28, 2001"
.SH NAME
tip \- connect to remote system
.SH SYNOPSIS
.LP
.nf
\fBtip\fR [\fB-v\fR] [\fB-speed-entry\fR] {\fIhostname\fR | \fIphone-number\fR | \fIdevice\fR}
.fi

.SH DESCRIPTION
.sp
.LP
The \fBtip\fR utility establishes a full-duplex terminal connection to a remote
host. Once the connection is established, a remote session using  \fBtip\fR
behaves like an interactive session on a local terminal.
.sp
.LP
The \fBremote\fR file contains entries describing remote systems and line
speeds used by \fBtip\fR.
.sp
.LP
Each host has a default baud rate for the connection, or you can specify a
speed with the \fB-\fR\fIspeed-entry\fR command line argument.
.sp
.LP
When \fIphone-number\fR is specified, \fBtip\fR looks for an entry in the
\fBremote\fR file of the form:
.sp
.in +2
.nf
tip -\fIspeed-entry\fR
.fi
.in -2
.sp

.sp
.LP
When \fBtip\fR finds such an entry, it sets the connection speed accordingly.
If it finds no such entry, \fBtip\fR interprets \fB-\fR\fIspeed-entry\fR as if
it were a system name, resulting in an error message.
.sp
.LP
If you omit \fB-\fR\fIspeed-entry\fR, \fBtip\fR uses the \fBtip0\fR entry to
set a speed for the connection.
.sp
.LP
When \fIdevice\fR is specified, \fBtip\fR attempts to open that device, but
will do so using the access privileges of the user, rather than \fBtip\fR's
usual access privileges (\fBsetuid uucp\fR). The user must have read/write
access to the device. The \fBtip\fR utility interprets any character string
beginning with the slash character (\|\fB/\fR\|) as a device name.
.sp
.LP
When establishing the connection, \fBtip\fR sends a connection message to the
remote system. The default value for this message can be found in the
\fBremote\fR file.
.sp
.LP
When \fBtip\fR attempts to connect to a remote system, it opens the associated
device with an exclusive-open \fBioctl\fR(2) call. Thus, only one user at a
time may access a device. This is to prevent multiple processes from sampling
the terminal line. In addition, \fBtip\fR honors the locking protocol used by
\fBuucp\fR(1C).
.sp
.LP
When \fBtip\fR starts up, it reads commands from the file \fB\&.tiprc\fR in
your home directory.
.SH OPTIONS
.sp
.ne 2
.na
\fB\fB-v\fR\fR
.ad
.RS 6n
Display commands from the \fB\&.tiprc\fR file as they are executed.
.RE

.SH USAGE
.sp
.LP
Typed characters are normally transmitted directly to the remote machine, which
does the echoing as well.
.sp
.LP
At any time that \fBtip\fR prompts for an argument (for example, during setup
of a file transfer), the line typed may be edited with the standard erase and
kill characters. A null line in response to a prompt, or an interrupt, aborts
the dialogue and returns you to the remote machine.
.SS "Commands"
.sp
.LP
A tilde (\fB~\fR) appearing as the first character of a line is an escape
signal which directs \fBtip\fR to perform some special action. \fBtip\fR
recognizes the following escape sequences:
.sp
.ne 2
.na
\fB\fB~^D\fR\fR
.ad
.br
.na
\fB\fB~.\fR\fR
.ad
.RS 18n
Drop the connection and exit (you may still be logged in on the remote
machine). \fBNote:\fR If you rlogin and then run \fBtip\fR on the remote host,
you must type \fB~~.\fR (tilde tilde dot) to end the \fBtip\fR session. If you
type \fB~.\fR (tilde dot), it terminates the rlogin.
.RE

.sp
.ne 2
.na
\fB\fB~c\fR [\fIname\fR]\fR
.ad
.RS 18n
Change directory to \fIname\fR. No argument implies change to your home
directory.
.RE

.sp
.ne 2
.na
\fB\fB~!\fR\fR
.ad
.RS 18n
Escape to an interactive shell on the local machine. Exiting the shell returns
you to \fBtip\fR.
.RE

.sp
.ne 2
.na
\fB\fB~>\fR\fR
.ad
.RS 18n
Copy file from local to remote.
.RE

.sp
.ne 2
.na
\fB\fB~<\fR\fR
.ad
.RS 18n
Copy file from remote to local.
.RE

.sp
.ne 2
.na
\fB\fB~p\fR\fI from\fR [\fI to \fR]\fR
.ad
.RS 18n
Send a file to a remote host running the UNIX system. When you use the put
command, the remote system runs the command string
.sp
.in +2
.nf
\fBcat > \fR\fIto\fR
.fi
.in -2
.sp

while \fBtip\fR sends it the \fIfrom\fR file. If the \fIto\fR file is not
specified, the \fIfrom\fR file name is used. This command is actually a
UNIX-system-specific version of the `\fB~>\fR' command.
.RE

.sp
.ne 2
.na
\fB\fB~t\fR\fI from\fR [\fI to \fR]\fR
.ad
.RS 18n
Take a file from a remote host running the UNIX system. As in the put command
the \fIto\fR file defaults to the \fIfrom\fR file name if it is not specified.
The remote host executes the command string
.sp
.in +2
.nf
\fBcat \fR\fIfrom\fR\fB\|;  echo ^A\fR
.fi
.in -2
.sp

to send the file to \fBtip\fR.
.RE

.sp
.ne 2
.na
\fB\fB~|\fR\fR
.ad
.RS 18n
Pipe the output from a remote command to a local process. The command string
sent to the local system is processed by the shell.
.RE

.sp
.ne 2
.na
\fB\fB~C\fR\fR
.ad
.RS 18n
Connect a program to the remote machine. The command string sent to the program
is processed by the shell. The program inherits file descriptors 0 as remote
line input, 1 as remote line output, and 2 as tty standard error.
.RE

.sp
.ne 2
.na
\fB\fB~$\fR\fR
.ad
.RS 18n
Pipe the output from a local process to the remote host. The command string
sent to the local system is processed by the shell.
.RE

.sp
.ne 2
.na
\fB\fB~#\fR\fR
.ad
.RS 18n
Send a \fBBREAK\fR to the remote system.
.RE

.sp
.ne 2
.na
\fB\fB~s\fR\fR
.ad
.RS 18n
Set a variable (see the discussion below).
.RE

.sp
.ne 2
.na
\fB\fB~^Z\fR\fR
.ad
.RS 18n
Stop \fBtip\fR. Only available when run under a shell that supports job
control, such as the C shell.
.RE

.sp
.ne 2
.na
\fB\fB~^Y\fR\fR
.ad
.RS 18n
Stop only the "local side" of \fBtip\fR. Only available when run under a shell
that supports job control, such as the C shell. The "remote side" of \fBtip\fR,
that is, the side that displays output from the remote host, is left running.
.RE

.sp
.ne 2
.na
\fB\fB~?\fR\fR
.ad
.RS 18n
Get a summary of the tilde escapes.
.RE

.sp
.LP
Copying files requires some cooperation on the part of the remote host. When a
\fB~>\fR or \fB~<\fR escape is used to send a file, \fBtip\fR prompts for a
file name (to be transmitted or received) and a command to be sent to the
remote system, in case the file is being transferred from the remote system.
While \fBtip\fR is transferring a file, the number of lines transferred will be
continuously displayed on the screen. A file transfer may be aborted with an
interrupt.
.SS "Auto-call Units"
.sp
.LP
\fBtip\fR may be used to dial up remote systems using a number of auto-call
unit's (\fBACU\fRs). When the remote system description contains the \fBdu\fR
capability, \fBtip\fR uses the call-unit (\fBcu\fR), \fBACU\fR type (\fBat\fR),
and phone numbers (\fBpn\fR) supplied. Normally, \fBtip\fR displays verbose
messages as it dials.
.sp
.LP
Depending on the type of auto-dialer being used to establish a connection, the
remote host may have garbage characters sent to it upon connection. The user
should never assume that the first characters typed to the foreign host are the
first ones presented to it. The recommended practice is to immediately type a
\fBkill\fR character upon establishing a connection (most UNIX systems either
support \fB@\fR or Control-U as the initial kill character).
.sp
.LP
\fBtip\fR currently supports the Ventel MD-212+ modem and DC Hayes-compatible
modems.
.sp
.LP
When \fBtip\fR initializes a Hayes-compatible modem for dialing, it sets up the
modem to auto-answer. Normally, after the conversation is complete, \fBtip\fR
drops \fBDTR\fR, which causes the modem to "hang up."
.sp
.LP
Most modems can be configured so that when \fBDTR\fR drops, they re-initialize
themselves to a preprogrammed state. This can be used to reset the modem and
disable auto-answer, if desired.
.sp
.LP
Additionally, it is possible to start the phone number with a Hayes \fBS\fR
command so that you can configure the modem before dialing. For example, to
disable auto-answer, set up all the phone numbers in \fB/etc/remote\fR using
something like \fBpn=S0=0DT5551212\fR. The  \fBS0=0\fR disables auto-answer.
.SS "Remote Host Description"
.sp
.LP
Descriptions of remote hosts are normally located in the system-wide file
\fB/etc/remote\fR. However, a user may maintain personal description files (and
phone numbers) by defining and exporting the \fBREMOTE\fR shell variable. The
\fBremote\fR file must be readable by \fBtip\fR, but a secondary file
describing phone numbers may be maintained readable only by the user. This
secondary phone number file is \fB/etc/phones\fR, unless the shell variable
\fBPHONES\fR is defined and exported. The phone number file contains lines of
the form:
.sp
.in +2
.nf
\fIsystem-name phone-number\fR
.fi
.in -2
.sp

.sp
.LP
Each phone number found for a system is tried until either a connection is
established, or an end of file is reached.  Phone numbers are constructed from
`\fB0123456789\(mi=*\fR', where the `\fB=\fR' and `\fB*\fR' are used to
indicate a second dial tone should be waited for (\fBACU\fR dependent).
.SS "tip Internal Variables"
.sp
.LP
\fBtip\fR maintains a set of variables which are used in normal operation. Some
of these variables are read-only to normal users (root is allowed to change
anything of interest). Variables may be displayed and set through the \fB~s\fR
escape.  The syntax for variables is patterned after \fBvi\fR(1) and
\fBmail\fR(1). Supplying \fBall\fR as an argument to the \fB~s\fR escape
displays all variables that the user can read.  Alternatively, the user may
request display of a particular variable by attaching a \fB?\fR to the end. For
example, `\fB~s escape?\fR' displays the current escape character.
.sp
.LP
Variables are numeric (num), string (str), character (char), or Boolean (bool)
values. Boolean variables are set merely by  specifying their name. They may be
reset by prepending a \fB!\fR to the name. Other variable types are set by
appending an \fB=\fR and the value. The entire assignment must not have any
blanks in it. A single set command may be used to interrogate as well as set a
number of variables.
.sp
.LP
Variables may be initialized at run time by placing set commands (without the
\fB~s\fR prefix) in a \fB\&.tiprc\fR file in one's home directory. The \fB-v\fR
option makes \fBtip\fR display the sets as they are made. Comments preceded by
a \fB#\fR sign can appear in the  \fB\&.tiprc\fR file.
.sp
.LP
Finally, the variable names must either be completely specified or an
abbreviation may be given.  The following list details those variables known to
\fBtip\fR.
.sp
.ne 2
.na
\fB\fBbeautify\fR\fR
.ad
.RS 16n
(bool) Discard unprintable characters when a session is being scripted;
abbreviated  \fBbe\fR. If the \fBnb\fR capability is present, \fBbeautify\fR is
initially set to \fBoff\fR. Otherwise, \fBbeautify\fR is initially set to
\fBon\fR.
.RE

.sp
.ne 2
.na
\fB\fBbaudrate\fR\fR
.ad
.RS 16n
(num) The baud rate at which the connection was established; abbreviated
\fBba\fR. If a baud rate was specified on the command line, \fBbaudrate\fR is
initially set to the specified value. Or, if the \fBbr\fR capability is
present, \fBbaudrate\fR is initially set to the value of that capability.
Otherwise, \fBbaudrate\fR is set to 300 baud. Once \fBtip\fR has been started,
\fBbaudrate\fR can only changed by the super-user.
.RE

.sp
.ne 2
.na
\fB\fBdialtimeout\fR\fR
.ad
.RS 16n
(num) When dialing a phone number, the time (in seconds) to wait for a
connection to be established; abbreviated \fBdial\fR. \fBdialtimeout\fR is
initially set to 60 seconds, and can only changed by the super-user.
.RE

.sp
.ne 2
.na
\fB\fBdisconnect\fR\fR
.ad
.RS 16n
(str) The string to send to the remote host to disconnect from it; abbreviated
\fBdi\fR. If the \fBdi\fR capability is present, \fBdisconnect\fR is initially
set to the value of that capability. Otherwise, \fBdisconnect\fR is set to a
null string (\fB""\fR).
.RE

.sp
.ne 2
.na
\fB\fBechocheck\fR\fR
.ad
.RS 16n
(bool) Synchronize with the remote host during file transfer by waiting for the
echo of the last character transmitted; abbreviated \fBec\fR. If the \fBec\fR
capability is present, \fBechocheck\fR is initially set to \fBon\fR. Otherwise,
\fBechocheck\fR is initially set to \fBoff\fR.
.RE

.sp
.ne 2
.na
\fB\fBeofread\fR\fR
.ad
.RS 16n
(str) The set of characters which signify an end-of-transmission during a
\fB~<\fR file transfer command; abbreviated \fBeofr\fR. If the \fBie\fR
capability is present, \fBeofread\fR is initially set to the value of that
capability. Otherwise, \fBeofread\fR is set to a null string (\fB""\fR).
.RE

.sp
.ne 2
.na
\fB\fBeofwrite\fR\fR
.ad
.RS 16n
(str) The string sent to indicate end-of-transmission during a \fB~>\fR file
transfer command; abbreviated \fBeofw\fR. If the \fBoe\fR capability is
present, \fBeofread\fR is initially set to the value of that capability.
Otherwise, \fBeofread\fR is set to a null string (\fB""\fR).
.RE

.sp
.ne 2
.na
\fB\fBeol\fR\fR
.ad
.RS 16n
(str) The set of characters which indicate an end-of-line. \fBtip\fR will
recognize escape characters only after an end-of-line. If the \fBel\fR
capability is present, \fBeol\fR is initially set to the value of that
capability. Otherwise, \fBeol\fR is set to a null string (\fB""\fR).
.RE

.sp
.ne 2
.na
\fB\fBescape\fR\fR
.ad
.RS 16n
(char) The command prefix (escape) character; abbreviated \fBes\fR. If the
\fBes\fR capability is present, \fBescape\fR is initially set to the value of
that capability. Otherwise, \fBescape\fR is set to `\fB\|~\|\fR\&'.
.RE

.sp
.ne 2
.na
\fB\fBetimeout\fR\fR
.ad
.RS 16n
(num) The amount of time, in seconds, that \fBtip\fR should wait for the
echo-check response when \fBechocheck\fR is set; abbreviated \fBet\fR. If the
\fBet\fR capability is present, \fBetimeout\fR is initially set to the value of
that capability. Otherwise, \fBetimeout\fR is set to 10 seconds.
.RE

.sp
.ne 2
.na
\fB\fBexceptions\fR\fR
.ad
.RS 16n
(str) The set of characters which should not be discarded due to the
beautification switch; abbreviated \fBex\fR. If the \fBex\fR capability is
present, \fBexceptions\fR is initially set to the value of that capability.
Otherwise, \fBexceptions\fR is set to `\fB\et\en\ef\eb\fR\&'.
.RE

.sp
.ne 2
.na
\fB\fBforce\fR\fR
.ad
.RS 16n
(char) The character used to force literal data transmission; abbreviated
\fBfo\fR. If the \fBfo\fR capability is present, \fBforce\fR is initially set
to the value of that capability. Otherwise, \fBforce\fR is set to \fB\e377\fR
(which disables it).
.RE

.sp
.ne 2
.na
\fB\fBframesize\fR\fR
.ad
.RS 16n
(num) The amount of data (in bytes) to buffer between file system writes when
receiving files; abbreviated \fBfr\fR. If the \fBfs\fR capability is present,
\fBframesize\fR is initially set to the value of that capability. Otherwise,
\fBframesize\fR is set to \fB1024\fR.
.RE

.sp
.ne 2
.na
\fB\fBhalfduplex\fR\fR
.ad
.RS 16n
(bool) Do local echoing because the host is half-duplex; abbreviated \fBhdx\fR.
If the \fBhd\fR capability is present, \fBhalfduplex\fR is initially set to
\fBon\fR. Otherwise, \fBhalfduplex\fR is initially set to \fBoff\fR.
.RE

.sp
.ne 2
.na
\fB\fBhardwareflow\fR\fR
.ad
.RS 16n
(bool) Do hardware flow control; abbreviated \fBhf\fR. If the  \fBhf\fR
capability is present, \fBhardwareflow\fR is initially set to \fBon\fR.
Otherwise, \fBhardwareflowcontrol\fR is initially set to \fBoff\fR.
.RE

.sp
.ne 2
.na
\fB\fBhost\fR\fR
.ad
.RS 16n
(str) The name of the host to which you are connected; abbreviated \fBho\fR.
\fBhost\fR is permanently set to the name given on the command line or in the
\fBHOST\fR environment variable.
.RE

.sp
.ne 2
.na
\fB\fBlocalecho\fR\fR
.ad
.RS 16n
(bool) A synonym for \fBhalfduplex\fR; abbreviated \fBle\fR.
.RE

.sp
.ne 2
.na
\fB\fBlog\fR\fR
.ad
.RS 16n
(str) The name of the file to which to log information about outgoing phone
calls. \fBlog\fR is initially set to \fB/var/adm/aculog\fR, and can only be
inspected or changed by the super-user.
.RE

.sp
.ne 2
.na
\fB\fBparity\fR\fR
.ad
.RS 16n
(str) The parity to be generated and checked when talking to the remote host;
abbreviated \fBpar\fR. The possible values are:
.sp
.ne 2
.na
\fB\fBnone\fR>\fR
.ad
.br
.na
\fB\fBzero\fR\fR
.ad
.RS 9n
Parity is not checked on input, and the parity bit is set to zero on output.
.RE

.sp
.ne 2
.na
\fB\fBone\fR\fR
.ad
.RS 9n
Parity is not checked on input, and the parity bit is set to one on output.
.RE

.sp
.ne 2
.na
\fB\fBeven\fR\fR
.ad
.RS 9n
Even parity is checked for on input and generated on output.
.RE

.sp
.ne 2
.na
\fB\fBodd\fR\fR
.ad
.RS 9n
Odd parity is checked for on input and generated on output.
.RE

If the \fBpa\fR capability is present, \fBparity\fR is initially set to the

usr/src/cmd/tip/tip.c
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

/*
 * tip - UNIX link to other systems
 *  tip [-v] [-speed] system-name
 * or
 *  cu phone-number [-s speed] [-l line] [-a acu]
 */
#include "tip.h"
#include <sys/wait.h>
#include <locale.h>

int	vflag;
int	cumode;
FILE	*phfd;

/*
 * Baud rate mapping table
 */
int bauds[] = {
	0, 50, 75, 110, 134, 150, 200, 300, 600,
	1200, 1800, 2400, 4800, 9600, 19200, 38400,
	57600, 76800, 115200, 153600, 230400, 307200,
	460800, 921600, 1000000, 1152000, 1500000,
	2000000, 2500000, 3000000, 3500000, 4000000, -1
};

extern void	tipout(void)	__NORETURN;
extern void	timeout(void);
extern esctable_t	etable[];
extern unsigned char	evenpartab[];

void	intprompt(void);
void	deadkid(void);
void	cleanup(void);
void	tipin(void)	__NORETURN;
unsigned char	escape(void);
char	*sname(char *);
char	PNbuf[256];		/* This limits the size of a number */
int	noparity = 0;

int
main(int argc, char *argv[])
{
	char *system = NOSTR;
	int i;
	char *p;
	char sbuf[15];

	gid = getgid();
	egid = getegid();
	uid = getuid();
	euid = geteuid();
	if (equal(sname(argv[0]), "cu")) {
		cumode = 1;
		cumain(argc, argv);
		goto cucommon;
	}

	if (argc > 4) {
		(void) fprintf(stderr,
		    "usage: tip [-v] [-speed] [system-name]\n");
		return (1);
	}
	if (!isatty(0)) {
		(void) fprintf(stderr, "tip: must be interactive\n");
		return (1);
	}

	for (; argc > 1; argv++, argc--) {
		if (argv[1][0] != '-')
			system = argv[1];
		else switch (argv[1][1]) {

		case 'v':
			vflag++;
			break;

		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			BR = atoi(&argv[1][1]);
			break;

		default:
			(void) fprintf(stderr, "tip: %s, unknown option\n",
			    argv[1]);
			break;
		}
	}

	(void) setlocale(LC_CTYPE, "");

	if (system == NOSTR)
		goto notnumber;
	for (p = system; *p; p++)
		if (isalpha(*p))
			goto notnumber;
	/*
	 * System name is really a phone number...
	 * Copy the number then stomp on the original (in case the number
	 *	is private, we don't want 'ps' or 'w' to find it).
	 */
	if (strlen(system) > sizeof (PNbuf) - 1) {
		(void) fprintf(stderr,
		    "tip: phone number too long (max = %d bytes)\n",
		    sizeof (PNbuf) - 1);
		return (1);
	}
	(void) strncpy(PNbuf, system, sizeof (PNbuf) - 1);
	for (p = system; *p; p++)
		*p = '\0';
	PN = PNbuf;
	(void) snprintf(sbuf, sizeof (sbuf), "tip%d", BR);
	system = sbuf;

notnumber:
	(void) signal(SIGINT, (sig_handler_t)cleanup);
	(void) signal(SIGQUIT, (sig_handler_t)cleanup);
	(void) signal(SIGHUP, (sig_handler_t)cleanup);
	(void) signal(SIGTERM, (sig_handler_t)cleanup);

	if ((i = hunt(system)) == 0) {
		(void) printf("all ports busy\n");
		return (3);
	}
	if (i == -1) {
		(void) printf("link down\n");
		delock(uucplock);
		return (3);
	}
	setbuf(stdout, NULL);
	loginit();

	/*
	 * Now that we have the logfile and the ACU open
	 *  return to the real uid and gid.  These things will
	 *  be closed on exit.  The saved-setuid uid and gid
	 *  allows us to get the original setuid permissions back
	 *  for removing the uucp lock.
	 */
	userperm();

	/*
	 * Kludge, there's no easy way to get the initialization
	 *   in the right order, so force it here.
	 * Do the open here, before we change back to real uid.
	 * We will check whether the open succeeded later, when
	 * (and if) we actually go to use the file.
	 */
	if ((PH = getenv("PHONES")) == NOSTR) {
		myperm();
		PH = "/etc/phones";
	}
	phfd = fopen(PH, "r");

	userperm();

	vinit();				/* init variables */
	setparity("none");			/* set the parity table */
	if ((i = speed(number(value(BAUDRATE)))) == 0) {
		(void) printf("tip: bad baud rate %d\n",
		    number(value(BAUDRATE)));
		myperm();
		delock(uucplock);
		return (3);
	}


	/*
	 * Hardwired connections require the
	 *  line speed set before they make any transmissions
	 *  (this is particularly true of things like a DF03-AC)
	 */
	if (HW)
		ttysetup(i);
	if (p = connect()) {
		(void) printf("\07%s\n[EOT]\n", p);
		myperm();
		delock(uucplock);
		return (1);
	}

	/*
	 * Always setup the tty again here in case hardware flow
	 *  control was selected, which can only be set after the
	 *  connection is made, or in case this is not a hardwired
	 *  modem (rare these days) that likewise can only be setup
	 *  after the connection is made.
	 */
	ttysetup(i);
cucommon:
	/*
	 * From here down the code is shared with
	 * the "cu" version of tip.
	 */

	(void) ioctl(0, TCGETS, (char *)&defarg);
	arg = defarg;
	/* turn off input processing */
	arg.c_lflag &= ~(ICANON|ISIG|ECHO|IEXTEN);
	arg.c_cc[VMIN] = 1;
	arg.c_cc[VTIME] = 0;
	arg.c_iflag &= ~(INPCK|IXON|IXOFF|ICRNL);
	arg.c_oflag = 0;		/* turn off all output processing */
	/* handle tandem mode in case was set in remote file */
	if (boolean(value(TAND)))
		tandem("on");
	else
		tandem("off");
	raw();

	(void) pipe(fildes); (void) pipe(repdes);
	(void) signal(SIGALRM, (sig_handler_t)timeout);

	/*
	 * Everything's set up now:
	 *	connection established (hardwired or dialup)
	 *	line conditioned (baud rate, mode, etc.)
	 *	internal data structures (variables)
	 * so, fork one process for local side and one for remote.
	 */
	if (CM != NOSTR) {
		(void) sleep(2);	/* let line settle */
		parwrite(FD, (unsigned char *)CM, strlen(CM));
	}
	(void) printf(cumode ? "Connected\r\n" : "\07connected\r\n");
	(void) signal(SIGCHLD, (sig_handler_t)deadkid);
	if (pid = fork())
		tipin();
	else
		tipout();
	/*NOTREACHED*/
}

void
deadkid(void)
{

	if (pid >= 0 && waitpid(pid, NULL, WNOHANG) == pid)
		tip_abort("Connection Closed");
}

void
cleanup(void)
{

	if (uid != getuid()) {
		myperm();
	}
	delock(uucplock);
	exit(0);
}

/*
 * put the controlling keyboard into raw mode
 */
void
raw(void)
{

	(void) ioctl(0, TCSETSF, (char *)&arg);
}


/*
 * return keyboard to normal mode
 */
void
unraw(void)
{

	(void) ioctl(0, TCSETSF, (char *)&defarg);
}

/*
 * switch to using invoking user's permissions
 */
void
userperm(void)
{

	(void) setegid(gid);
	(void) seteuid(uid);
}

/*
 * switch to using my special (setuid) permissions
 */
void
myperm(void)
{

	(void) setegid(egid);
	(void) seteuid(euid);
}

static	sigjmp_buf promptbuf;

/*
 * Print string ``s'', then read a string
 *  in from the terminal.  Handles signals & allows use of
 *  normal erase and kill characters.
 */
int
prompt(char *s, char *p, size_t len)
{
	char *b = p;
	int c;
	sig_handler_t	ointr, oquit;

	stoprompt = 0;
	ointr = signal(SIGINT, (sig_handler_t)intprompt);
	oquit = signal(SIGQUIT, SIG_IGN);
	unraw();
	(void) printf("%s", s);
	if (sigsetjmp(promptbuf, 1) == 0)
		while (p < b + len - 1 &&
		    ((c = getchar()) != EOF) && (c != '\n'))
			*p++ = c;
	*p = '\0';

	raw();
	(void) signal(SIGINT, ointr);
	(void) signal(SIGQUIT, oquit);
	return (stoprompt || p == b);
}

/*
 * Interrupt service routine during prompting
 */
void
intprompt(void)
{

	(void) signal(SIGINT, SIG_IGN);
	(void) signal(SIGQUIT, SIG_IGN);
	stoprompt = 1;
	(void) printf("\r\n");
	siglongjmp(promptbuf, 1);
}

/*
 * ****TIPIN   TIPIN****
 */
void
tipin(void)
{
	unsigned char gch, c;
	int bol = 1;

	/*
	 * Kinda klugey here...
	 *   check for scripting being turned on from the .tiprc file,
	 *   but be careful about just using setscript(), as we may
	 *   send a SIGEMT before tipout has a chance to set up catching
	 *   it; so wait a second, then setscript()
	 */
	if (boolean(value(SCRIPT))) {
		(void) sleep(1);
		setscript();
	}

	for (;;) {
		gch = getchar()&0377;
		if ((gch == character(value(ESCAPE))) && bol) {
			if (!(gch = escape()))
				continue;
		} else if (!cumode && gch == character(value(RAISECHAR))) {
			boolean(value(RAISE)) = !boolean(value(RAISE));
			continue;
		} else if (gch == '\r') {
			bol = 1;
			parwrite(FD, &gch, 1);
			if (boolean(value(HALFDUPLEX)))
				(void) printf("\r\n");
			continue;
		} else if (!cumode && gch == character(value(FORCE)))
			gch = getchar()&0377;
		bol = any(gch, value(EOL));
		if (boolean(value(RAISE)) && islower(gch))
			gch = toupper(gch);
		c = gch;
		parwrite(FD, &gch, 1);
		if (boolean(value(HALFDUPLEX)))
			(void) putchar(c);
	}
}

/*
 * Escape handler --
 *  called on recognition of ``escapec'' at the beginning of a line
 */
unsigned char
escape(void)
{
	unsigned char gch;
	esctable_t *p;
	char c = character(value(ESCAPE));

	gch = (getchar()&0377);
	for (p = etable; p->e_char; p++)
		if (p->e_char == gch) {
			if ((p->e_flags&PRIV) && uid)
				continue;
			(void) printf("%s", ctrl(c));
			(*p->e_func)(gch);
			return (0);
		}
	/* ESCAPE ESCAPE forces ESCAPE */
	if (c != gch)
		parwrite(FD, (unsigned char *)&c, 1);
	return (gch);
}

int
speed(int n)
{
	int *p;

	for (p = bauds; *p != -1;  p++)
		if (*p == n)
			return (p - bauds);
	return (0);
}

int
any(char c, char *p)
{
	while (p && *p)
		if (*p++ == c)
			return (1);
	return (0);
}

char *
interp(char *s)
{
	static char buf[256];
	char *p = buf, c, *q;

	while (c = *s++) {
		for (q = "\nn\rr\tt\ff\033E\bb"; *q; q++)
			if (*q++ == c) {
				*p++ = '\\'; *p++ = *q;
				goto next;
			}
		if (c < 040) {
			*p++ = '^'; *p++ = c + 'A'-1;
		} else if (c == 0177) {
			*p++ = '^'; *p++ = '?';
		} else
			*p++ = c;
	next:
		;
	}
	*p = '\0';
	return (buf);
}

char *
ctrl(char c)
{
	static char s[3];

	if (c < 040 || c == 0177) {
		s[0] = '^';
		s[1] = c == 0177 ? '?' : c+'A'-1;
		s[2] = '\0';
	} else {
		s[0] = c;
		s[1] = '\0';
	}
	return (s);
}

/*
 * Help command
 */
void
help(int c)
{
	esctable_t *p;

	(void) printf("%c\r\n", c);
	for (p = etable; p->e_char; p++) {
		if ((p->e_flags&PRIV) && uid)
			continue;
		(void) printf("%2s", ctrl(character(value(ESCAPE))));
		(void) printf("%-2s %c   %s\r\n", ctrl(p->e_char),
		    p->e_flags&EXP ? '*': ' ', p->e_help);
	}
}

/*
 * Set up the "remote" tty's state
 */
void
ttysetup(int speed)
{
	struct termios buf;
	char *loc;

	(void) ioctl(FD, TCGETS, (char *)&buf);
	buf.c_cflag &= (CREAD|HUPCL|CLOCAL|CRTSCTS|CRTSXOFF);
	buf.c_cflag |= CS8;
	(void) cfsetospeed(&buf, speed);
	if (boolean(value(HARDWAREFLOW))) {
		int i = TIOCM_CAR;

		/*
		 * Only set hardware flow control if carrier is up,
		 * because some devices require both CD and RTS to
		 * be up before sending.
		 */
		(void) ioctl(FD, TIOCMGET, &i);
		if (i & TIOCM_CAR)
			buf.c_cflag |= (CRTSCTS|CRTSXOFF);
	}

	/*
	 * Careful to only penalize the 8-bit users here on the
	 * incoming tty port.  The default 7-bit users will
	 * still get the parity bit from the other side's login
	 * process (which happens to be the default for sun tip
	 * configurations).
	 */
	loc = setlocale(LC_CTYPE, NULL);
	if (noparity && loc != 0 && strcmp(loc, "C") != 0)
		buf.c_iflag = 0;
	else
		buf.c_iflag = ISTRIP;
	buf.c_oflag = 0;
	buf.c_lflag = 0;
	buf.c_cc[VMIN] = 1;
	buf.c_cc[VTIME] = 0;
	(void) ioctl(FD, TCSETSF, (char *)&buf);
}

/*
 * Return "simple" name from a file name,
 * strip leading directories.
 */
char *
sname(char *s)
{
	char *p = s;

	while (*s)
		if (*s++ == '/')
			p = s;
	return (p);
}

static char partab[0400];

/*
 * Do a write to the remote machine with the correct parity.
 * We are doing 8 bit wide output, so we just generate a character
 * with the right parity and output it.
 */
void
parwrite(int fd, unsigned char *buf, int n)
{
	int i;
	unsigned char *bp;

	bp = buf;
	for (i = 0; i < n; i++) {
		*bp = partab[(*bp)&0377];
		bp++;
	}
	if (write(fd, buf, n) < 0) {
		if (errno == EIO || errno == ENXIO)
			tip_abort("Lost carrier.");
		/* this is questionable */
		perror("write");
	}
}

/*
 * Build a parity table with appropriate high-order bit.
 */
void
setparity(char *defparity)
{
	int i;
	char *parity;

	if (value(PARITY) == NOSTR)
		value(PARITY) = defparity;
	parity = value(PARITY);
	for (i = 0; i < 0400; i++)
		partab[i] = evenpartab[i];
	if (equal(parity, "even")) {
		/* EMPTY */
	} else if (equal(parity, "odd")) {
		for (i = 0; i < 0400; i++)
			partab[i] ^= 0200;	/* reverse bit 7 */
	} else if (equal(parity, "none")) {
		/* Do nothing so we can pass thru 8-bit chars */
		noparity = 1;
		for (i = 0; i < 0400; i++)
			partab[i] = i;
	} else if (equal(parity, "zero")) {
		for (i = 0; i < 0400; i++)
			partab[i] &= ~0200;	/* turn off bit 7 */
	} else if (equal(parity, "one")) {
		for (i = 0; i < 0400; i++)
			partab[i] |= 0200;	/* turn on bit 7 */
	} else {
		(void) fprintf(stderr, "%s: unknown parity value\n", PA);
		(void) fflush(stderr);
	}
}
